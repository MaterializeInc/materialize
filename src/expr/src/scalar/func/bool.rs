// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Boolean functions.

use repr::strconv;
use repr::{Datum, RowArena, ScalarType};

use crate::scalar::func::{FuncProps, Nulls, OutputType};
use crate::scalar::{EvalError, ScalarExpr};

pub const CAST_BOOL_TO_STRING_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: true,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::String),
};

pub fn cast_bool_to_string_explicit(a: Datum) -> Result<Datum, EvalError> {
    // N.B. this function differs from `cast_bool_to_string_implicit` because
    // the SQL specification requires `true` and `false` to be spelled out
    // in explicit casts, while PostgreSQL prefers its more concise `t` and `f`
    // representation in implicit casts.
    match a.unwrap_bool() {
        true => Ok(Datum::from("true")),
        false => Ok(Datum::from("false")),
    }
}

pub fn cast_bool_to_string_implicit(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::String(strconv::format_bool_static(a.unwrap_bool())))
}

pub const CAST_STRING_TO_BOOL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn cast_string_to_bool(a: Datum) -> Result<Datum, EvalError> {
    match strconv::parse_bool(a.unwrap_str())? {
        true => Ok(Datum::True),
        false => Ok(Datum::False),
    }
}

pub const AND_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: false,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn and<'a>(
    datums: &[Datum<'a>],
    temp_storage: &'a RowArena,
    a_expr: &'a ScalarExpr,
    b_expr: &'a ScalarExpr,
) -> Result<Datum<'a>, EvalError> {
    match a_expr.eval(datums, temp_storage)? {
        Datum::False => Ok(Datum::False),
        a => match (a, b_expr.eval(datums, temp_storage)?) {
            (_, Datum::False) => Ok(Datum::False),
            (Datum::Null, _) | (_, Datum::Null) => Ok(Datum::Null),
            (Datum::True, Datum::True) => Ok(Datum::True),
            _ => unreachable!(),
        },
    }
}

pub const OR_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: false,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn or<'a>(
    datums: &[Datum<'a>],
    temp_storage: &'a RowArena,
    a_expr: &'a ScalarExpr,
    b_expr: &'a ScalarExpr,
) -> Result<Datum<'a>, EvalError> {
    match a_expr.eval(datums, temp_storage)? {
        Datum::True => Ok(Datum::True),
        a => match (a, b_expr.eval(datums, temp_storage)?) {
            (_, Datum::True) => Ok(Datum::True),
            (Datum::Null, _) | (_, Datum::Null) => Ok(Datum::Null),
            (Datum::False, Datum::False) => Ok(Datum::False),
            _ => unreachable!(),
        },
    }
}

pub const NOT_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: true,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn not(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(!a.unwrap_bool()))
}
