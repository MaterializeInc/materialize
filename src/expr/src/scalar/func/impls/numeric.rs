// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use dec::{OrderedDecimal, Rounding};
use repr::adt::numeric::{self, Numeric};
use repr::strconv;

use crate::EvalError;

sqlfunc!(
    #[sqlname = "-"]
    #[preserves_uniqueness = true]
    fn neg_numeric(mut a: Numeric) -> Numeric {
        numeric::cx_datum().neg(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

sqlfunc!(
    #[sqlname = "abs"]
    fn abs_numeric(mut a: Numeric) -> Numeric {
        numeric::cx_datum().abs(&mut a);
        a
    }
);

sqlfunc!(
    #[sqlname = "ceilnumeric"]
    fn ceil_numeric(mut a: Numeric) -> Numeric {
        // ceil will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        let mut cx = numeric::cx_datum();
        cx.set_rounding(Rounding::Ceiling);
        cx.round(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

sqlfunc!(
    #[sqlname = "expnumeric"]
    fn exp_numeric(mut a: Numeric) -> Result<Numeric, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.exp(&mut a);
        let cx_status = cx.status();
        if cx_status.overflow() {
            Err(EvalError::FloatOverflow)
        } else if cx_status.subnormal() {
            Err(EvalError::FloatUnderflow)
        } else {
            numeric::munge_numeric(&mut a).unwrap();
            Ok(a)
        }
    }
);

sqlfunc!(
    #[sqlname = "floornumeric"]
    fn floor_numeric(mut a: Numeric) -> Numeric {
        // floor will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        let mut cx = numeric::cx_datum();
        cx.set_rounding(Rounding::Floor);
        cx.round(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        a
    }
);

fn log_guard_numeric(val: &Numeric, function_name: &str) -> Result<(), EvalError> {
    if val.is_negative() {
        return Err(EvalError::NegativeOutOfDomain(function_name.to_owned()));
    }
    if val.is_zero() {
        return Err(EvalError::ZeroOutOfDomain(function_name.to_owned()));
    }
    Ok(())
}

// From the `decNumber` library's documentation:
// > Inexact results will almost always be correctly rounded, but may be up to 1
// > ulp (unit in last place) in error in rare cases.
//
// See decNumberLog10 documentation at http://speleotrove.com/decimal/dnnumb.html
fn log_numeric<F>(mut a: Numeric, logic: F, name: &'static str) -> Result<Numeric, EvalError>
where
    F: Fn(&mut dec::Context<Numeric>, &mut Numeric),
{
    log_guard_numeric(&a, name)?;
    let mut cx = numeric::cx_datum();
    logic(&mut cx, &mut a);
    numeric::munge_numeric(&mut a).unwrap();
    Ok(a)
}

sqlfunc!(
    #[sqlname = "lnnumeric"]
    fn ln_numeric(a: Numeric) -> Result<Numeric, EvalError> {
        log_numeric(a, dec::Context::ln, "ln")
    }
);

sqlfunc!(
    #[sqlname = "log10numeric"]
    fn log10_numeric(a: Numeric) -> Result<Numeric, EvalError> {
        log_numeric(a, dec::Context::log10, "log10")
    }
);

sqlfunc!(
    #[sqlname = "roundnumeric"]
    fn round_numeric(mut a: Numeric) -> Numeric {
        // round will be nop if has no fractional digits.
        if a.exponent() >= 0 {
            return a;
        }
        numeric::cx_datum().round(&mut a);
        a
    }
);

sqlfunc!(
    #[sqlname = "sqrtnumeric"]
    fn sqrt_numeric(mut a: Numeric) -> Result<Numeric, EvalError> {
        if a.is_negative() {
            return Err(EvalError::NegSqrt);
        }
        let mut cx = numeric::cx_datum();
        cx.sqrt(&mut a);
        numeric::munge_numeric(&mut a).unwrap();
        Ok(a)
    }
);

sqlfunc!(
    #[sqlname = "numerictoi16"]
    fn cast_numeric_to_int16(mut a: Numeric) -> Result<i16, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        let i = cx.try_into_i32(a).or(Err(EvalError::Int16OutOfRange))?;
        i16::try_from(i).or(Err(EvalError::Int16OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictoi32"]
    fn cast_numeric_to_int32(mut a: Numeric) -> Result<i32, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        cx.try_into_i32(a).or(Err(EvalError::Int32OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictoi64"]
    fn cast_numeric_to_int64(mut a: Numeric) -> Result<i64, EvalError> {
        let mut cx = numeric::cx_datum();
        cx.round(&mut a);
        cx.clear_status();
        cx.try_into_i64(a).or(Err(EvalError::Int64OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "numerictof32"]
    fn cast_numeric_to_float32(a: Numeric) -> Result<f32, EvalError> {
        let i = a.to_string().parse::<f32>().unwrap();
        if i.is_infinite() {
            Err(EvalError::Float32OutOfRange)
        } else {
            Ok(i)
        }
    }
);

sqlfunc!(
    #[sqlname = "numerictof64"]
    fn cast_numeric_to_float64(a: Numeric) -> Result<f64, EvalError> {
        let i = a.to_string().parse::<f64>().unwrap();
        if i.is_infinite() {
            Err(EvalError::Float64OutOfRange)
        } else {
            Ok(i)
        }
    }
);

sqlfunc!(
    #[sqlname = "numerictostr"]
    fn cast_numeric_to_string(a: Numeric) -> String {
        let mut buf = String::new();
        strconv::format_numeric(&mut buf, &OrderedDecimal(a));
        buf
    }
);
