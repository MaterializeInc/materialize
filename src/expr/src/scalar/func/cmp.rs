// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Comparison functions.

use repr::{Datum, ScalarType};

use crate::scalar::func::{FuncProps, Nulls, OutputType};
use crate::scalar::EvalError;

pub const CMP_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Bool),
};

pub fn eq<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a == b))
}

pub fn not_eq<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a != b))
}

pub fn lt<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a < b))
}

pub fn lte<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a <= b))
}

pub fn gt<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a > b))
}

pub fn gte<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a >= b))
}
