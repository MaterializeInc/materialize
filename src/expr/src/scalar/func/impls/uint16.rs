// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::adt::numeric::{self, Numeric, NumericMaxScale};
use mz_repr::{strconv, ColumnType, ScalarType};

use crate::scalar::func::EagerUnaryFunc;
use crate::EvalError;

sqlfunc!(
    #[sqlname = "~"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::BitNotUint16)]
    fn bit_not_uint16(a: u16) -> u16 {
        !a
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_real"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastFloat32ToUint16)]
    fn cast_uint16_to_float32(a: u16) -> f32 {
        f32::from(a)
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_double"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastFloat64ToUint16)]
    fn cast_uint16_to_float64(a: u16) -> f64 {
        f64::from(a)
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_uint4"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastUint32ToUint16)]
    fn cast_uint16_to_uint32(a: u16) -> u32 {
        u32::from(a)
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_uint8"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastUint64ToUint16)]
    fn cast_uint16_to_uint64(a: u16) -> u64 {
        u64::from(a)
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_smallint"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastInt16ToUint16)]
    fn cast_uint16_to_int16(a: u16) -> Result<i16, EvalError> {
        i16::try_from(a).or(Err(EvalError::Int16OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_integer"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastInt32ToUint16)]
    fn cast_uint16_to_int32(a: u16) -> i32 {
        i32::from(a)
    }
);
sqlfunc!(
    #[sqlname = "uint2_to_bigint"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastInt64ToUint16)]
    fn cast_uint16_to_int64(a: u16) -> i64 {
        i64::from(a)
    }
);

sqlfunc!(
    #[sqlname = "uint2_to_text"]
    #[preserves_uniqueness = true]
    #[right_inverse = to_unary!(super::CastStringToUint16)]
    fn cast_uint16_to_string(a: u16) -> String {
        let mut buf = String::new();
        strconv::format_uint16(&mut buf, a);
        buf
    }
);

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct CastUint16ToNumeric(pub Option<NumericMaxScale>);

impl<'a> EagerUnaryFunc<'a> for CastUint16ToNumeric {
    type Input = u16;
    type Output = Result<Numeric, EvalError>;

    fn call(&self, a: u16) -> Result<Numeric, EvalError> {
        let mut a = Numeric::from(i32::from(a));
        if let Some(scale) = self.0 {
            if numeric::rescale(&mut a, scale.into_u8()).is_err() {
                return Err(EvalError::NumericFieldOverflow);
            }
        }
        Ok(a)
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Numeric { max_scale: self.0 }.nullable(input.nullable)
    }

    fn right_inverse(&self) -> Option<crate::UnaryFunc> {
        to_unary!(super::CastNumericToUint16)
    }
}

impl fmt::Display for CastUint16ToNumeric {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("uint2_to_numeric")
    }
}
