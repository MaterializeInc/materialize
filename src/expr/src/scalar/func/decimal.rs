// Copyright Materialize, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Decimal functions.

use std::cmp::Ordering;

use ore::collections::CollectionExt;
use ore::result::ResultExt;
use repr::adt::decimal::MAX_DECIMAL_PRECISION;
use repr::{strconv, Datum, RowArena, ScalarType};

use crate::scalar::func::{float, FuncProps, Nulls, OutputType};
use crate::scalar::EvalError;

pub const CAST_DECIMAL_TO_STRING_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::String),
};

pub fn cast_decimal_to_string<'a>(
    a: Datum<'a>,
    scale: u8,
    temp_storage: &'a RowArena,
) -> Result<Datum<'a>, EvalError> {
    let mut buf = String::new();
    strconv::format_decimal(&mut buf, &a.unwrap_decimal().with_scale(scale));
    Ok(Datum::String(temp_storage.push_string(buf)))
}

pub fn cast_string_to_decimal_props(scale: u8) -> FuncProps {
    FuncProps {
        can_error: true,
        preserves_uniqueness: false,
        nulls: Nulls::Sometimes {
            propagates_nulls: true,
            introduces_nulls: false,
        },
        output_type: OutputType::Fixed(ScalarType::Decimal(MAX_DECIMAL_PRECISION, scale)),
    }
}

pub fn cast_string_to_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    strconv::parse_decimal(a.unwrap_str())
        .map(|d| {
            Datum::from(match d.scale().cmp(&scale) {
                Ordering::Less => d.significand() * 10_i128.pow(u32::from(scale - d.scale())),
                Ordering::Equal => d.significand(),
                Ordering::Greater => d.significand() / 10_i128.pow(u32::from(d.scale() - scale)),
            })
        })
        .err_into()
}

pub const CAST_DECIMAL_TO_INT32_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Int32),
};

pub fn cast_decimal_to_int32(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as i32))
}

pub const CAST_DECIMAL_TO_INT64_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Int64),
};

pub fn cast_decimal_to_int64(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as i64))
}

pub const CAST_DECIMAL_TO_FLOAT32_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Float32),
};

pub fn cast_decimal_to_float32(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as f32))
}

pub const CAST_DECIMAL_TO_FLOAT64_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Fixed(ScalarType::Float64),
};

pub fn cast_decimal_to_float64(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().as_i128() as f64))
}

pub const DECIMAL_MATH_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::MatchesInput,
};

pub fn abs_decimal(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(a.unwrap_decimal().abs()))
}

pub fn neg_decimal(a: Datum) -> Result<Datum, EvalError> {
    Ok(Datum::from(-a.unwrap_decimal()))
}

pub fn add_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a.unwrap_decimal() + b.unwrap_decimal()))
}

pub fn sub_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a.unwrap_decimal() - b.unwrap_decimal()))
}

pub fn mod_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    let b = b.unwrap_decimal();
    if b == 0 {
        Err(EvalError::DivisionByZero)
    } else {
        Ok(Datum::from(a.unwrap_decimal() % b))
    }
}

pub const MUL_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Computed(|input_types| {
        assert_eq!(input_types.len(), 2);
        let (_, s1) = input_types[0].scalar_type.unwrap_decimal_parts();
        let (_, s2) = input_types[1].scalar_type.unwrap_decimal_parts();
        ScalarType::Decimal(MAX_DECIMAL_PRECISION, s1 + s2)
    }),
};

pub fn mul_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    Ok(Datum::from(a.unwrap_decimal() * b.unwrap_decimal()))
}

pub const DIV_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Computed(|input_types| {
        assert_eq!(input_types.len(), 2);
        let (_, s1) = input_types[0].scalar_type.unwrap_decimal_parts();
        let (_, s2) = input_types[1].scalar_type.unwrap_decimal_parts();
        ScalarType::Decimal(MAX_DECIMAL_PRECISION, s1 - s2)
    }),
};

pub fn div_decimal<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError> {
    let b = b.unwrap_decimal();
    if b == 0 {
        Err(EvalError::DivisionByZero)
    } else {
        Ok(Datum::from(a.unwrap_decimal() / b))
    }
}

pub fn ceil_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    let decimal = a.unwrap_decimal();
    Ok(Datum::from(decimal.with_scale(scale).ceil().significand()))
}

pub fn floor_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    let decimal = a.unwrap_decimal();
    Ok(Datum::from(decimal.with_scale(scale).floor().significand()))
}

pub const ROUND_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: false,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::Computed(|input_types| input_types.into_first().scalar_type),
};

pub fn round_decimal_unary(a: Datum, a_scale: u8) -> Result<Datum, EvalError> {
    round_decimal_binary(a, Datum::Int64(0), a_scale)
}

pub fn round_decimal_binary<'a>(
    a: Datum<'a>,
    b: Datum<'a>,
    a_scale: u8,
) -> Result<Datum<'a>, EvalError> {
    let round_to = b.unwrap_int64();
    let decimal = a.unwrap_decimal().with_scale(a_scale);
    Ok(Datum::from(decimal.round(round_to).significand()))
}

pub const SQRT_DECIMAL_PROPS: FuncProps = FuncProps {
    can_error: true,
    preserves_uniqueness: false,
    nulls: Nulls::Sometimes {
        propagates_nulls: true,
        introduces_nulls: false,
    },
    output_type: OutputType::MatchesInput,
};

pub fn sqrt_decimal(a: Datum, scale: u8) -> Result<Datum, EvalError> {
    let d = a.unwrap_decimal();
    if d.as_i128() < 0 {
        return Err(EvalError::NegSqrt);
    }
    let d_f64 = cast_decimal_to_float64(a)?;
    let d_scaled = d_f64.unwrap_float64() / 10_f64.powi(i32::from(scale));
    float::cast_float64_to_decimal(Datum::from(d_scaled.sqrt()), scale)
}
