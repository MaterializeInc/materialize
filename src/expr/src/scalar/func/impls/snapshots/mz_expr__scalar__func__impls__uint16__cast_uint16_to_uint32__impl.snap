---
source: src/expr/src/scalar/func/impls/uint16.rs
expression: info
---
Info {
    output_type_nullable: ColumnType {
        scalar_type: UInt32,
        nullable: true,
    },
    output_type_nonnullable: ColumnType {
        scalar_type: UInt32,
        nullable: false,
    },
    preserves_uniqueness: true,
    inverse: Some(
        CastUint32ToUint16(
            CastUint32ToUint16,
        ),
    ),
    is_monotone: true,
    propagates_nulls: true,
    introduces_nulls: false,
    could_error: false,
}
